### ISSUE TITLE
Kernel version upgrade performed by unattended-upgrades breaks FSx client.

### ISSUE THREAD

**Issue by asford** [aws/aws-parallelcluster#2095]
**Required Info:**
 - AWS ParallelCluster version [e.g. 2.9.0]: 2.9.1
 - Cluster name: main
 - Config below, our bootstrap script inserts lustre mounts via `/etc/fstab`.

**Bug description and how to reproduce:**

The EC2 instances of a fresh parallelcluster installation on ubuntu1804 come up with kernel version 5.4.0-1035-aws. If the machine reboots the kernel version changes to 5.4.0-1025-aws, as reported by `uname -r`.

 This breaks the any fsx mounts, as the 1035 lustre-client-modules-aws are still installed. A subsequent call to `apt get install lustre-client-modules-aws` fixes this issue by installing the correct lustre client modules, however I'm puzzled as to why the kernel version shift is occurring in the first place.

This can be trivially reproducing by creating a pcluster with an `fsx` configuration, then calling `sudo reboot` on the head node. The filesystem will fail to mount on reboot.

### Config

```config
# AUTOGENERATED BY `terraform apply`

[aws]
aws_region_name          = ca-central-1

[global]
cluster_template         = default
update_check             = false
sanity_check             = true

[vpc default]
vpc_id                   = vpc-0b0815bf6e400689c
master_subnet_id         = subnet-02e0c4b67b3dd2898
compute_subnet_id        = subnet-09eb59ef851b89868

vpc_security_group_id    = sg-0a6f0d6ff3f1dd2c6
additional_sg            = sg-0d0f8c19104396f03

[cluster default]
vpc_settings             = default
key_name                 = abcbio-spml_admin_key_20200928
scheduler                = slurm
queue_settings           = m5,p3
base_os                  = ubuntu1804
master_instance_type     = m5.xlarge
additional_iam_policies  = arn:aws:iam::074555333423:policy/pcluster_compute_bucket_access
post_install             = s3://internal-pcluster-spml-abcbio-ca-central-1/pcluster/bootstrap
post_install_args        = 'setup s3://internal-pcluster-spml-abcbio-ca-central-1/pcluster'




[queue m5]
compute_resource_settings    = m5
enable_efa                   = false
disable_hyperthreading       = true
compute_type                 = ondemand


[compute_resource m5]
instance_type                = m5.8xlarge
min_count                    = 0
initial_count                = 0
max_count                    = 25


[queue p3]
compute_resource_settings    = p3
enable_efa                   = false
disable_hyperthreading       = true
compute_type                 = ondemand

placement_group              = DYNAMIC


[compute_resource p3]
instance_type                = p3.2xlarge
min_count                    = 0
initial_count                = 0
max_count                    = 8
```

**Comment by asford** [comment]
I've verified a repo on my end and confirmed the version numbers. You are correct, the node comes up with `5.3.0-1035-aws` and updates to `5.4.0-1025-aws` on first boot. This is due to an upgrade of some form that isn't triggering the corresponding lustre component updates. This is present in `/var/log/apt/history.log` as a call from `unattended-upgrade`. 

I can provide more logs if it's useful, but would appreciate your guidance on where to look next as I'm not fully familiar with the pcluster image creation / cfn-init process.

The repro log is below, lightly edited to remove extraneous components for clarity.

### Cluster create, before reboot, 5.4.0-1035-aws kernel and lustre
```bash

$ pcluster create test-2095
Beginning cluster creation for cluster: test-2095
Creating stack named: parallelcluster-test-2095
Status: parallelcluster-test-2095 - CREATE_COMPLETE
MasterPublicIP: 52.60.43.140
ClusterUser: ubuntu
MasterPrivateIP: 172.31.57.134

$ pcluster status test-2095
Status: CREATE_COMPLETE
MasterServer: RUNNING
MasterPublicIP: 52.60.43.140
ClusterUser: ubuntu
MasterPrivateIP: 172.31.57.134
ComputeFleetStatus: RUNNING


$ ssh ubuntu@52.60.43.140
The authenticity of host '52.60.43.140 (52.60.43.140)' can't be established.
ECDSA key fingerprint is SHA256:ydf94L9I/mIrUD1uKd7gznV+MTyKCymK8WrAv4SRaoY.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '52.60.43.140' (ECDSA) to the list of known hosts.
Welcome to Ubuntu 18.04.5 LTS (GNU/Linux 5.3.0-1035-aws x86_64)
[...]

36 packages can be updated.
26 updates are security updates.


*** System restart required ***
Last login: Mon Sep 14 23:05:29 2020 from 54.174.110.255

ubuntu@ip-172-31-57-134:~$ uname -r
5.3.0-1035-aws

$ cat /var/log/apt/history.log

[...]

Start-Date: 2020-10-01  17:00:31
Commandline: /usr/bin/unattended-upgrade
Install: linux-aws-5.4-headers-5.4.0-1025:amd64 (5.4.0-1025.25~18.04.1, automatic), linux-image-5.4.0-1025-aws:amd64 (5.4.0-1025.25~18.04.1, automatic), linux-modules-5.4.0-1025-aws:amd64 (5.4.0-1025.25~18.04.1, automatic), linux-headers-5.4.0-1025-aws:amd64 (5.4.0-1025.25~18.04.1, automatic)
Upgrade: linux-headers-aws:amd64 (5.3.0.1035.34, 5.4.0.1025.10), linux-aws:amd64 (5.3.0.1035.34, 5.4.0.1025.10), linux-image-aws:amd64 (5.3.0.1035.34, 5.4.0.1025.10)
End-Date: 2020-10-01  17:02:32

Start-Date: 2020-10-01  17:02:34
Commandline: /usr/bin/unattended-upgrade
Install: linux-headers-4.15.0-118-generic:amd64 (4.15.0-118.119, automatic), linux-modules-4.15.0-118-generic:amd64 (4.15.0-118.119, automatic), linux-image-4.15.0-118-generic:amd64 (4.15.0-118.119, automatic), linux-headers-4.15.0-118:amd64 (4.15.0-118.119, automatic), linux-modules-extra-4.15.0-118-generic:amd64 (4.15.0-118.119, automatic)
Upgrade: linux-headers-generic:amd64 (4.15.0.117.104, 4.15.0.118.105), linux-image-generic:amd64 (4.15.0.117.104, 4.15.0.118.105), linux-generic:amd64 (4.15.0.117.104, 4.15.0.118.105)
End-Date: 2020-10-01  17:05:21kj

[...]

ubuntu@ip-172-31-57-134:~$ sudo reboot
Connection to 52.60.43.140 closed by remote host.
Connection to 52.60.43.140 closed.
```

### Post-reboot, 5.4.0-1025-aws kernel and  5.4.0-1035-aws lustre

```
$ ssh ubuntu@52.60.43.140
Welcome to Ubuntu 18.04.5 LTS (GNU/Linux 5.4.0-1025-aws x86_64)
[...]
10 packages can be updated.
0 updates are security updates.

New release '20.04.1 LTS' available.
Run 'do-release-upgrade' to upgrade to it.

Last login: Thu Oct  1 17:14:59 2020 from 208.98.202.66

ubuntu@ip-172-31-57-134:~$ uname -r
5.4.0-1025-aws


ubuntu@ip-172-31-57-134:~$ sudo apt list | grep lustre

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

lustre-client-modules-4.15.0-1054-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1056-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1057-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1058-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1060-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1061-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1062-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1063-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1065-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1066-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-4.15.0-1067-aws/unknown 2.10.8-1fsx3 amd64
lustre-client-modules-4.15.0-1073-aws/unknown 2.10.8-1fsx3 amd64
lustre-client-modules-4.15.0-1077-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-4.15.0-1079-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-4.15.0-1080-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1017-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-5.3.0-1019-aws/unknown 2.10.8-1fsx1 amd64
lustre-client-modules-5.3.0-1020-aws/unknown 2.10.8-1fsx4 amd64
lustre-client-modules-5.3.0-1023-aws/unknown 2.10.8-1fsx4 amd64
lustre-client-modules-5.3.0-1028-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1030-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1032-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1033-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1034-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-5.3.0-1035-aws/unknown,now 2.10.8-1fsx5 amd64 [installed]
lustre-client-modules-5.4.0-1025-aws/unknown 2.10.8-1fsx5 amd64
lustre-client-modules-aws/unknown 5.4.0.1025.10 amd64 [upgradable from: 5.3.0.1035.34]
lustre-dev/unknown 2.10.8-1fsx5 amd64
lustre-iokit/unknown 2.10.8-1fsx5 amd64
lustre-source/unknown 2.10.8-1fsx5 all
lustre-tests/unknown 2.10.8-1fsx5 amd64
lustre-utils/unknown,now 2.10.8-1fsx5 amd64 [installed,automatic]
```
[test_2095.txt](https://github.com/aws/aws-parallelcluster/files/5313802/test_2095.txt)

**Comment by demartinofra** [comment]
Hi Alex,

here is an explanation of the behavior you are experiencing:

The official Ubuntu 18.04 AMI we use as the starting point to build ParallelCluster AMIs has unattended upgrades (https://help.ubuntu.com/community/AutomaticSecurityUpdates) enabled to run once per day. 

```
ubuntu@ip-10-0-0-55:~$ cat /etc/apt/apt.conf.d/20auto-upgrades
APT::Periodic::Update-Package-Lists "1";
APT::Periodic::Unattended-Upgrade "1";
```

While this daily processes takes care of pulling security updates and upgrading the kernel version, the Lustre client package update is skipped due to the default configuration of the allowed origins for unattended upgrades. 

```
ubuntu@ip-10-0-0-55:~$ cat /etc/apt/apt.conf.d/50unattended-upgrades
// Automatically upgrade packages from these (origin:archive) pairs
//
// Note that in Ubuntu security updates may pull in new dependencies
// from non-security sources (e.g. chromium). By allowing the release
// pocket these get automatically pulled in.
Unattended-Upgrade::Allowed-Origins {
    "${distro_id}:${distro_codename}";
    "${distro_id}:${distro_codename}-security";
    // Extended Security Maintenance; doesn't necessarily exist for
    // every release and this system may not have it installed, but if
    // available, the policy for updates is such that unattended-upgrades
    // should also install from here by default.
    "${distro_id}ESMApps:${distro_codename}-apps-security";
    "${distro_id}ESM:${distro_codename}-infra-security";
//    "${distro_id}:${distro_codename}-updates";
//    "${distro_id}:${distro_codename}-proposed";
//    "${distro_id}:${distro_codename}-backports";
};

...
```

Because of this, if a user reboots the instance after unattended upgrades have pulled a newer kernel version, the kernel gets upgraded while the `lustre-client-modules-aws` is kept to an older incompatible version.
Here is a dry-run output of the unattended-upgrades program, showing that `lustre-client-modules-aws` is ignored:

```
ubuntu@ip-10-0-0-55:~$ sudo unattended-upgrades --dry-run --debug
Initial blacklisted packages:
Initial whitelisted packages:
Starting unattended upgrades script
Allowed origins are: o=Ubuntu,a=bionic, o=Ubuntu,a=bionic-security, o=UbuntuESMApps,a=bionic-apps-security, o=UbuntuESM,a=bionic-infra-security
Using (^linux-image-[0-9]+\.[0-9\.]+-.*|^linux-headers-[0-9]+\.[0-9\.]+-.*|^linux-image-extra-[0-9]+\.[0-9\.]+-.*|^linux-modules-[0-9]+\.[0-9\.]+-.*|^linux-modules-extra-[0-9]+\.[0-9\.]+-.*|^linux-signed-image-[0-9]+\.[0-9\.]+-.*|^linux-image-unsigned-[0-9]+\.[0-9\.]+-.*|^kfreebsd-image-[0-9]+\.[0-9\.]+-.*|^kfreebsd-headers-[0-9]+\.[0-9\.]+-.*|^gnumach-image-[0-9]+\.[0-9\.]+-.*|^.*-modules-[0-9]+\.[0-9\.]+-.*|^.*-kernel-[0-9]+\.[0-9\.]+-.*|^linux-backports-modules-.*-[0-9]+\.[0-9\.]+-.*|^linux-modules-.*-[0-9]+\.[0-9\.]+-.*|^linux-tools-[0-9]+\.[0-9\.]+-.*|^linux-cloud-tools-[0-9]+\.[0-9\.]+-.*|^linux-buildinfo-[0-9]+\.[0-9\.]+-.*|^linux-source-[0-9]+\.[0-9\.]+-.*|^linux-image-[0-9]+\.[0-9\.]+-.*|^linux-headers-[0-9]+\.[0-9\.]+-.*|^linux-image-extra-[0-9]+\.[0-9\.]+-.*|^linux-modules-[0-9]+\.[0-9\.]+-.*|^linux-modules-extra-[0-9]+\.[0-9\.]+-.*|^linux-signed-image-[0-9]+\.[0-9\.]+-.*|^linux-image-unsigned-[0-9]+\.[0-9\.]+-.*|^kfreebsd-image-[0-9]+\.[0-9\.]+-.*|^kfreebsd-headers-[0-9]+\.[0-9\.]+-.*|^gnumach-image-[0-9]+\.[0-9\.]+-.*|^.*-modules-[0-9]+\.[0-9\.]+-.*|^.*-kernel-[0-9]+\.[0-9\.]+-.*|^linux-backports-modules-.*-[0-9]+\.[0-9\.]+-.*|^linux-modules-.*-[0-9]+\.[0-9\.]+-.*|^linux-tools-[0-9]+\.[0-9\.]+-.*|^linux-cloud-tools-[0-9]+\.[0-9\.]+-.*|^linux-buildinfo-[0-9]+\.[0-9\.]+-.*|^linux-source-[0-9]+\.[0-9\.]+-.*) regexp to find kernel packages
Using (^linux-image-5\.3\.0\-1035\-aws$|^linux-headers-5\.3\.0\-1035\-aws$|^linux-image-extra-5\.3\.0\-1035\-aws$|^linux-modules-5\.3\.0\-1035\-aws$|^linux-modules-extra-5\.3\.0\-1035\-aws$|^linux-signed-image-5\.3\.0\-1035\-aws$|^linux-image-unsigned-5\.3\.0\-1035\-aws$|^kfreebsd-image-5\.3\.0\-1035\-aws$|^kfreebsd-headers-5\.3\.0\-1035\-aws$|^gnumach-image-5\.3\.0\-1035\-aws$|^.*-modules-5\.3\.0\-1035\-aws$|^.*-kernel-5\.3\.0\-1035\-aws$|^linux-backports-modules-.*-5\.3\.0\-1035\-aws$|^linux-modules-.*-5\.3\.0\-1035\-aws$|^linux-tools-5\.3\.0\-1035\-aws$|^linux-cloud-tools-5\.3\.0\-1035\-aws$|^linux-buildinfo-5\.3\.0\-1035\-aws$|^linux-source-5\.3\.0\-1035\-aws$|^linux-image-5\.3\.0\-1035\-aws$|^linux-headers-5\.3\.0\-1035\-aws$|^linux-image-extra-5\.3\.0\-1035\-aws$|^linux-modules-5\.3\.0\-1035\-aws$|^linux-modules-extra-5\.3\.0\-1035\-aws$|^linux-signed-image-5\.3\.0\-1035\-aws$|^linux-image-unsigned-5\.3\.0\-1035\-aws$|^kfreebsd-image-5\.3\.0\-1035\-aws$|^kfreebsd-headers-5\.3\.0\-1035\-aws$|^gnumach-image-5\.3\.0\-1035\-aws$|^.*-modules-5\.3\.0\-1035\-aws$|^.*-kernel-5\.3\.0\-1035\-aws$|^linux-backports-modules-.*-5\.3\.0\-1035\-aws$|^linux-modules-.*-5\.3\.0\-1035\-aws$|^linux-tools-5\.3\.0\-1035\-aws$|^linux-cloud-tools-5\.3\.0\-1035\-aws$|^linux-buildinfo-5\.3\.0\-1035\-aws$|^linux-source-5\.3\.0\-1035\-aws$) regexp to find running kernel packages
Checking: cloud-init ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: cloud-init=18.2-14-g6d48d265-0ubuntu1
Checking: initramfs-tools ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: initramfs-tools=0.130ubuntu3.6
Checking: initramfs-tools-bin ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: initramfs-tools-bin=0.130ubuntu3.6
Checking: initramfs-tools-core ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: initramfs-tools-core=0.130ubuntu3.6
Checking: libsane-common ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: libsane-common=1.0.27-1~experimental3ubuntu2.3
Checking: libsane1 ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: libsane1=1.0.27-1~experimental3ubuntu2.3
Checking: lustre-client-modules-aws ([<Origin component:'main' archive:'' origin:'' label:'' site:'fsx-lustre-client-repo.s3.amazonaws.com' isTrusted:True>])
Checking: sane-utils ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: sane-utils=1.0.27-1~experimental3ubuntu2.3
Checking: snapd ([<Origin component:'main' archive:'bionic-updates' origin:'Ubuntu' label:'Ubuntu' site:'us-east-2.ec2.archive.ubuntu.com' isTrusted:True>])
adjusting candidate version: snapd=2.45.1+18.04.2
pkgs that look like they should be upgraded:
Fetched 0 B in 0s (0 B/s)
fetch.run() result: 0
blacklist: []
whitelist: []
Removing unused kernel packages: linux-image-4.15.0-117-generic linux-modules-4.15.0-117-generic linux-modules-extra-4.15.0-117-generic
marking linux-image-4.15.0-117-generic for removal
marking linux-modules-4.15.0-117-generic for removal
marking linux-modules-extra-4.15.0-117-generic for removal
Packages that were successfully auto-removed:
Packages that are kept back:
No packages found that can be upgraded unattended and no pending auto-removals
``` 

This is due to the fact that the allowed origins configuration does not allow repos with an empty `origin` and an empty `archive` field values (`Checking: lustre-client-modules-aws ([<Origin component:'main' archive:'' origin:'' label:'' site:'fsx-lustre-client-repo.s3.amazonaws.com' isTrusted:True>]`).
Adding ":"; to the Unattended-Upgrade::Allowed-Origins makes it so that unattended upgrades process also tries to update the Lustre client. The downside is that ":"; risks to whitelist additional packages and not only Lustre client.
Unfortunately the fsx-lustre-client-repo does not set any specific value for the optional `origin` and `archive` fields which would have allowed a more granular configuration of the `Unattended-Upgrade::Allowed-Origins`.

We will follow-up on this issue to understand how to properly configure unattended upgrades in order to have FSx client upgraded together with the kernel.

**Comment by mpatton125** [comment]
It appears as though this issue has been addressed in version 5.4.1037, as the repo now includes an origin and suite:

"AWS-FSx-for-Lustre:stable"

**Comment by rexcsn** [comment]
Hi, [AWS ParallelCluster v2.10.2](https://github.com/aws/aws-parallelcluster/releases/tag/v2.10.2) is released and contains fix for this issue.
Closing the issue now, please feel free to reopen if there is any question. Thank you!
